game-app/game-core/src/main/java/games/strategy/triplea/settings/lobby/LobbySelectionView.java

```diff
@@ -0,0 +1,145 @@
+import javax.swing.AbstractButton;
+import javax.swing.ButtonGroup;
+import javax.swing.JComponent;
+import javax.swing.JRadioButton;
+import javax.swing.JTextField;
+import org.triplea.java.StringUtils;
+import org.triplea.swing.JTextFieldBuilder;
+import org.triplea.swing.SwingComponents;
+import org.triplea.swing.jpanel.JPanelBuilder;
+import org.triplea.ui.events.queue.ControllerListener;
+import org.triplea.ui.events.queue.ViewClass;
+
+/**
+ * UI with: <br>
+ * 1) Radio buttons to select lobby, eg: prod / prerelease / local / other <br>
+ * 2) A text field for custom input of lobby URI, enabled only if 'other' is selected.<br>
+ *
+ * <p>The lobby text field is pre-filled and disabled for each radio button choice except for
+ * 'other' which will enable the uri field.
+ */
+class LobbySelectionView implements SelectionComponent<JComponent>, ViewClass {
+
+  private final LobbySelectionEventQueue lobbySelectionEventQueue;
+  private final JComponent uiComponent;
+  private final JTextField uriField;
+  private final List<JRadioButton> radioButtons;
+
+  LobbySelectionView(final LobbySelectionEventQueue lobbySelectionEventQueue) {
+    this.lobbySelectionEventQueue = lobbySelectionEventQueue;
+    final LobbySelectionViewData uiData = new LobbySelectionViewData();
+
+    uriField = new JTextFieldBuilder().build();
+    uriField.setText(uiData.getUriFieldValue());
+    uriField.setEnabled(uiData.getUriFieldEnabled());
+
+    final ButtonGroup buttonGroup = new ButtonGroup();
+    radioButtons =
+        // create a radio button for each of the lobby selection options
+        Stream.of(LobbySelectionViewData.LobbyChoice.values())
+            .map(
+                choice -> {
+                  final JRadioButton radioButton = new JRadioButton(choice.getDisplayString());
+                  buttonGroup.add(radioButton);
+                  radioButton.addActionListener(
+                      e ->
+                          lobbySelectionEventQueue.publishUiEvent(
+                              LobbySelectionEvents.ViewEvent.RADIO_BUTTON_CHANGED,
+                              readDataState()));
+
+                  return radioButton;
+                })
+            .collect(Collectors.toList());
+
+    uiComponent =
+        new JPanelBuilder()
+            .boxLayoutVertical()
+            .add(new JPanelBuilder().addAll(new ArrayList<>(radioButtons)).build())
+            .add(uriField)
+            .build();
+    updateUiToState(new LobbySelectionViewData());
+  }
+
+  @SuppressWarnings("SwitchStatementWithTooFewBranches")
+  @ControllerListener
+  public void updateUiToState(
+      final LobbySelectionEvents.ControllerEvent controllerEvent,
+      final UnaryOperator<LobbySelectionViewData> dataChangeFunction) {
+
+    switch (controllerEvent) {
+      case UPDATE_UI_DATA:
+        final var updatedState = dataChangeFunction.apply(readDataState());
+        updateUiToState(updatedState);
+        break;
+    }
+  }
```

Comment:
Slightly off-topic, but I find it fascinating to see reccuring programming pattern regardless of the language being used.
This is basically reactive programming and reminds me of my job where currently most of my time goes into JavaScript programming using react, and this is basically how data is handled there: You define how data should be rendered and then you only update this data source using different approaches (the react-redux framework does the exact same, by "publishing"/"dispatching" global events to modify the data store, whereas the default react framework prefers a more direct approach where there isn't as much abstraction).

Reply:
The design here is very much meant to be a reactive-style.

I was thinking a nice future modification could be to have an automatic progress dialog appear for events that are slow to be completed. We could have an 'acknowledge' event method to remove an event off a queue. A dialog watcher could check the for queue for items that are over 500ms old and show a dialog (or a status message).
 
Stepping back a bit, I think a reactive event-queue style with view-controller-view-model is a good combination. The before and after seem like pretty good improvements. Adding another radio button turned out to be surprisingly complex, in this updated style it seemed straight forward (switch statements and the simplified data model combine to make it pretty easy to write out the logic cases).

@RoiEXLab I don't know much about React. I'm curious about the 'more direct approach', is there an easy example of how this update could be changed to have that direct approach? Would it be a good thing?

Reply:
I'll try my best to explain it without going too much into the details of react:
So basically almost everything you code with react is a "Component" and using so-called hooks the code looks something like this:
```js
function MyComponent() {
  return <div>
      <span>My Text here</span>
      <button>Some Button</button>
    </div>;
}
```

which would render pretty much exactly the HTML you can see there in the browser. If you want to have some dynamic data in there using react only without any additional frameworks you would do it with a so called "state" that can be modified by your code.

```js
function MyComponent() {
  const [ text, setText ] = React.useState('My Text here');
  const handleClick = event => setText('Updated Text');

  return <div>
      <span>{text}</span>
      <button onClick={handleClick}>Some Button</button>
    </div>;
}
```

So whenever `text` is modified via `setText` react will notice that and re-render the components that did change (so in this case `MyComponent`, `div` and `span` will be re-rendered IIRC, the `button` will be kept around and simply re-attached to the DOM).
And this is pretty much what I meant with a more direct approach to data storage. Basically you define that you need to store a single value (can be more complex but you need to keep in mind to keep it simple enough to not re-render unneccesarily often, just because some single property changed there). If you wanted to modify this value elsewhere, you'd have to define it in the common parent, and pass the setter to the child methods that need access, there is a way to have some sort of global storage, but it rarely makes sense to use it.

Using redux, the framework I mentioned previously the data storage becomes global and the component would looks something like this:
```js
// heavily simplified for clarity
function MyComponentContainer() {
  return <MyComponent  text={store.myComponentText} setText={text => dispatch(SET_TEXT_EVENT, text)}/>;
}

const myComponentReducer(event, currentState) {
  switch(event.action) {
    case SET_TEXT_EVENT:
      return {
        ...currentState,
        myComponentText: event.payload,
      };
  }
}

function MyComponent(props) {
  const { text, setText } = props;
  const handleClick = event => setText('Updated Text');

  return <div>
      <span>{text}</span>
      <button onClick={handleClick}>Some Button</button>
    </div>;
}
```

Which means that all instances of `MyComponent` would share the same underlying storage which can be desirable in some cases but in other cases this is not what you want.

So I don't know if a more direct approach suits our problem in java better, it would require us to implement a generic data wrapper that notifies all listeners whenever it is updated, but it will probably not reach the same level of syntactic sugar as in react.

If you have any questions or I explained something badly feel free to ask.

Reply:
Interesting, thanks @RoiEXLab .
It sounds like the approach taken for the PBF screen was more of the direct approach (maybe?). In that design every element had an event listener attached to it that then updated the view-model data state. Then any action clicks were a no-arg request to the view-model to execute the action.

Just for contrast, the event-queue design presented here has the view emitting an immutable view-model object with an event type all in one go and only when 'action clicks' occur. The PBF model did seem to run into performance problems and IMO the thread-hopping was not as clean.

I'm thinking that this event queue model can be extended further and for more contexts. It certainly is 'reactive' programming inspired. To that extent I was thinking the next addition could be for automatic progress dialogs. To do that a queue event would have an optional 'completed callback' attached to it. Our event queue would then check for events that are still in queue (not completed) and if still in progress after 300ms then we show a progress dialog (though we also really need to make our current progress dialog to be closeable)